app/src/main/java/org/projectbuendia/client/ui/OdkActivityLauncher.java

```diff
@@ -396,118 +426,277 @@ public static void sendOdkResultToServer(
-                    LOG.e(error, "Did not submit form to server successfully");
-
-                    SubmitXformFailedEvent.Reason reason =
-                        SubmitXformFailedEvent.Reason.UNKNOWN;
-                    if (error.networkResponse != null) {
-                        switch (error.networkResponse.statusCode) {
-                            case 401:
-                            case 403:
-                                reason = SubmitXformFailedEvent.Reason.SERVER_AUTH;
-                                break;
-                            case 404:
-                                reason = SubmitXformFailedEvent.Reason.SERVER_BAD_ENDPOINT;
-                                break;
-                            case 500:
-                                if (error.networkResponse.data == null) {
-                                    LOG.e("Server error, but no internal error stack trace "
-                                        + "available.");
-                                } else {
-                                    LOG.e(new String(
-                                        error.networkResponse.data, Charsets.UTF_8));
-                                    LOG.e("Server error. Internal error stack trace:\n");
-                                }
-                                reason = SubmitXformFailedEvent.Reason.SERVER_ERROR;
-                                break;
-                            default:
-                                reason = SubmitXformFailedEvent.Reason.SERVER_ERROR;
-                                break;
-                        }
+        connection.postXformInstance(patientUuid, xml, successListener, errorListener);
+    }
+
+    /** Tries to submit to the server all unsent forms. Returns {@code true} if there are no more
+     * unsent forms. Otherwise returns {@code false}.
+     */
+    public static final boolean resendFormsToServer(final ContentResolver contentResolver) {
+        final boolean hasUnsubmittedForms[] = new boolean[]{false};
+        for(final UnsyncForm unsyncForm : getUnsyncForms(contentResolver)) {
+            sendFormToServer(unsyncForm.patientUuid, unsyncForm.xml,
+                new Response.Listener<JSONObject>() {
+                    @Override public void onResponse(JSONObject response) {
+                        LOG.i("Created new encounter successfully on server. " + response.toString());
+                        deleteUnsyncForm(unsyncForm.uuid, contentResolver);
+
                     }
+                }, new Response.ErrorListener() {
+                    @Override public void onErrorResponse(VolleyError error) {
+                        //Just log it and flag as not synchronized, this form is already persisted.
+                        LOG.e(error, format("Error resubmitting %s form to server ", unsyncForm.uuid));
+                        hasUnsubmittedForms[0] = true;
+                    }
+                });
+        }
+        return !hasUnsubmittedForms[0];
```

Comment:
Sure. Take a look at this Gist, I've written a couple of examples for you: https://gist.github.com/capnfabs/d9e8dac58d23ac89c838

Normally, if you just had one thing you were submitting, you would define a `Future`, set the Future in the callbacks, and then wait on the future before returning. The idea with `Futures` is that it represents a value that will be available in the future, but isn't yet. I've demonstrated this in `WaitForSingleForm.java`.

Because we've got multiple forms instead, we can wait for all of them to complete by using a `CountDownLatch`. I've included an example for this in the gist, the example file is called `OdkActivityLauncher.java`. `CountDownLatch`es block until they've counted down to 0, so in this example, we create a CountDownLatch with a size equal to the number of requests we're going to send, and then when each request either completes or fails, we count down. In the `resendFormsToServerBlocking` method, we wait until the CountDownLatch has counted down completely, and then return the value of `hasUnsubmittedForms`, which we know will be up to date.

Does that make sense?